### PR TITLE
Improve sparql testing

### PR DESCRIPTION
**Pull request by brinxmat**
- Again, a lot of code here — mostly POJO-ish
- Adds framework for testing sparql queries and responses
- Codacy fails because the branch is not analyzed.

### CODE REVIEW

File: report-api/src/main/resources/template/affiliation.sparql [hunk 1]
```diff
@@ -2,53 +2,59 @@ PREFIX : <https://nva.sikt.no/ontology/publication#>
...
             }
         }
     }
 
-    FILTER(LANG(?label) = "nb")
+    BIND(STR(?affiliationUri) as ?affiliationId)
+    BIND(STR(?institutionUri) as ?institutionId)
+    BIND(STR(?facultyUri) as ?facultyId)
+    BIND(STR(?departmentUri) as ?departmentId)
+    BIND(STR(?groupUri) as ?groupId)
```
Comment (brinxmat) [inline report-api/src/main/resources/template/affiliation.sparql:50]: Ensure response is a String, not a URI

File: report-api/src/main/resources/template/affiliation.sparql [hunk 2]
```diff
@@ -2,53 +2,59 @@ PREFIX : <https://nva.sikt.no/ontology/publication#>
...
+
+    FILTER(LANG(?label) = "en")
     BIND(STR(?label) AS ?affiliationName)
 
     ?personId :name ?contributorName .
-    BIND(REPLACE(STR(?personId), "https://api.nva.unit.no/cristin/person/", "", "i") AS ?contributorId)
+    BIND(REPLACE(STR(?personId), "^.*/([^/]*)$", "$1") AS ?contributorId)
```
Comment (brinxmat) [inline report-api/src/main/resources/template/affiliation.sparql:56]: We just want the last path element

File: report-api/src/main/resources/template/affiliation.sparql [hunk 3]
```diff
@@ -2,53 +2,59 @@ PREFIX : <https://nva.sikt.no/ontology/publication#>
...
     FILTER(!ISBLANK(?personId))
     FILTER(?modifiedDate >= "__AFTER__"^^xsd:dateTime)
     FILTER(?modifiedDate < "__BEFORE__"^^xsd:dateTime)
-} LIMIT __PAGE_SIZE__ OFFSET __OFFSET__
+} ORDER BY ?publicationId LIMIT __PAGE_SIZE__ OFFSET __OFFSET__
```
Comment (brinxmat) [inline report-api/src/main/resources/template/affiliation.sparql:60]: Ensure a deterministic ordering

File: report-api/src/test/java/no/sikt/nva/data/report/api/fetch/FetchDataReportTest.java [hunk 4]
```diff
@@ -45,29 +62,39 @@ void shouldThrowBadRequest(TestingRequest report)
 
     @ParameterizedTest
     @ArgumentsSource(ValidRequestSource.class)
-    void shouldReturnFormattedResult(TestingRequest request) throws IOException {
-        var service = new QueryService(setupDatabaseConnection());
+    void shouldReturnFormattedResult(TestingRequest request) throws IOException, BadRequestException {
+        var testData = new TestData(List.of(new DatePair(new PublicationDate("2023", "02", "02"),
+                                                     Instant.now().minus(100, ChronoUnit.DAYS)),
+                                        new DatePair(new PublicationDate("2023", "10", "18"),
+                                                     Instant.now().minus(100, ChronoUnit.DAYS))));
```
Comment (brinxmat) [inline report-api/src/test/java/no/sikt/nva/data/report/api/fetch/FetchDataReportTest.java:67]: This is the crux of the issue, we can now generate full records with differences that allow us to filter them.

File: report-api/src/test/java/no/sikt/nva/data/report/api/fetch/FetchDataReportTest.java [hunk 5]
```diff
@@ -45,29 +62,39 @@ void shouldThrowBadRequest(TestingRequest report)
...
+    private String getExpected(TestingRequest request, TestData test) throws BadRequestException {
+        var responseType = TEXT_CSV.toString().equals(request.acceptHeader().get(ACCEPT))
+                               ? TEXT_CSV
+                               : TEXT_PLAIN;
+        var data = switch (ReportType.parse(request.reportType())) {
+            case AFFILIATION -> test.getAffiliationResponseData();
+            case CONTRIBUTOR -> test.getContributorResponseData();
+            case FUNDING -> test.getFundingResponseData();
+            case IDENTIFIER -> test.getIdentifierResponseData();
+            case PUBLICATION -> test.getPublicationResponseData();
```
Comment (brinxmat) [inline report-api/src/test/java/no/sikt/nva/data/report/api/fetch/FetchDataReportTest.java:90]: The expected result can be calculated from the generated model. The formatting is also generated from the model.

File: report-api/src/test/java/no/sikt/nva/data/report/api/fetch/db/FakeDatabaseConnection.java [hunk 6]
```diff
@@ -25,6 +26,20 @@ public String getResult(Query query, ResponseFormatter formatter) {
...
+
+    public void insert(Model data) {
+        model.add(data);
+    }
+
+    public String dump() {
+        var stringWriter = new StringWriter();
+        RDFDataMgr.write(stringWriter, model, Lang.TURTLE);
+        return stringWriter.toString();
+    }
```
Comment (brinxmat) [inline report-api/src/test/java/no/sikt/nva/data/report/api/fetch/db/FakeDatabaseConnection.java:37]: This is a helper method for debugging

File: report-api/src/test/java/no/sikt/nva/data/report/api/fetch/formatter/ExpectedCsvFormatter.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+import com.opencsv.CSVReader;
+import com.opencsv.exceptions.CsvException;
+import java.io.IOException;
+import java.io.StringReader;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class ExpectedCsvFormatter {
```
Comment (brinxmat) [inline report-api/src/test/java/no/sikt/nva/data/report/api/fetch/formatter/ExpectedCsvFormatter.java:12]: This method formats CSV input to the format returned by the Jena formatter — the code is deceptively simple.

File: report-api/src/test/java/no/sikt/nva/data/report/api/fetch/formatter/ExpectedCsvFormatter.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    private static String printRow(String[] row, int[] maxLengths, boolean isHeader) {
+        return IntStream.range(0, row.length)
+                   .mapToObj(i -> formatCell(i, row, maxLengths, isHeader))
+                   .collect(Collectors.joining()) + COLUMN_SEPARATOR;
+    }
+
+    private static String formatCell(int index, String[] row, int[] maxLengths, boolean isHeader) {
+        var cell = row[index];
+        var value = isHeader || cell.isBlank() ? cell : String.format(QUOTED_STRING_TEMPLATE, cell);
+        return String.format("| %-" + maxLengths[index] + "s ", value);
```
Comment (brinxmat) [inline report-api/src/test/java/no/sikt/nva/data/report/api/fetch/formatter/ExpectedCsvFormatter.java:81]: The String::format template with maxlength provides the padding

File: report-api/src/test/java/no/sikt/nva/data/report/api/fetch/testutils/generator/TestChannel.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    public PublicationContext toModel() {
+        return type.contains("Journal")
+                   ? new JournalGenerator(identifier)
+                         .withName(name)
+                         .withScientificValue(scientificValue)
+                         .withOnlineIssn(onlineIssn)
+                         .withPrintIssn(printIssn)
+                   : new PublisherGenerator(identifier)
+                         .withName(name)
+                         .withScientificValue(scientificValue);
```
Comment (brinxmat) [inline report-api/src/test/java/no/sikt/nva/data/report/api/fetch/testutils/generator/TestChannel.java:60]: This method is common to all Test* model classes, it allows us to generate the necessary RDF to populate the database.

File: report-api/src/test/java/no/sikt/nva/data/report/api/fetch/testutils/generator/TestOrganization.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+            throw new RuntimeException("The last path element in the organization URI should be formatted as 10.0.0.0");
+        }
+        if (parts[1].equals("0")) {
+            return null;
+        } else if (parts[2].equals("0")) {
+            return new TestOrganization(baseUri + parts[0] + ".0.0.0", null);
+        } else if (parts[3].equals("0")) {
+            return new TestOrganization(baseUri + parts[0] + "." + parts[1] + ".0.0", null);
+        } else {
+            return new TestOrganization(baseUri + parts[0] + "." + parts[1] + "." + parts[2] + ".0", null);
```
Comment (brinxmat) [inline report-api/src/test/java/no/sikt/nva/data/report/api/fetch/testutils/generator/TestOrganization.java:56]: The code is designed to produce a recognisable organisation structure to match what we see in the real-world the data.

File: report-api/src/test/java/no/sikt/nva/data/report/api/fetch/testutils/generator/TestChannel.java [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+        return this;
+    }
+
+    public TestChannel withScientificValue(String value) {
+        this.scientificValue = value;
+        return this;
+    }
+
+    public PublicationContext toModel() {
+        return type.contains("Journal")
```
[the remark this answers is not in the document]
Comment (monaullah) [reply report-api/src/test/java/no/sikt/nva/data/report/api/fetch/testutils/generator/TestChannel.java:52]: (Not sure if it is relevant)

**Review by monaullah — APPROVED** [review]

File: report-api/src/test/java/no/sikt/nva/data/report/api/fetch/testutils/generator/TestChannel.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (brinxmat) [reply report-api/src/test/java/no/sikt/nva/data/report/api/fetch/testutils/generator/TestChannel.java:52]: Probably something that needs to be defined as relevant before I do anything with it @ketilaa

**Pull request merged**